nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py

```diff
@@ -217,34 +221,41 @@ def _process_example(self, example):
             elif self.truncation_field == "context":
                 context_ids = context_ids[: -min(truncation_length, len(context_ids))]
 
-        if len(context_ids) > self.max_seq_length:
-            context_ids = context_ids[: self.max_seq_length]
-
-        assert len(context_ids) <= self.max_seq_length
         input_ids = context_ids
-
         answer_start_idx = len(input_ids)
+
+        # Adds bos token in the start
+        if self.add_bos:
+            context_ids = [self.tokenizer.bos_id] + context_ids
+            input_ids = [self.tokenizer.bos_id] + input_ids
+            answer_start_idx += 1
+
         # Adds sep token between text/prompt and answer
         if self.add_sep:
+            context_ids = context_ids + [self.sep_id]
             input_ids = input_ids + [self.sep_id]
             answer_start_idx += 1
 
         input_ids = input_ids + answer_ids
 
-        if self.add_bos:
-            input_ids = [self.tokenizer.bos_id] + input_ids
-            answer_start_idx += 1
-        if self.add_eos:
+        # Only training need to consider eos token
+        if self.add_eos and self.tokens_to_generate == 0:
             input_ids = input_ids + [self.tokenizer.eos_id]
 
-        if len(input_ids) < self.min_seq_length or len(input_ids) > self.max_seq_length:
+        if len(input_ids) > self.max_seq_length:
+            logging.warning(f'Input ids length {len(input_ids)} exceed max sequence length {self.max_seq_length}')
             input_ids = input_ids[: self.max_seq_length]
 
+        # store metadata in dataset, in case user may have keys required in the prediction json files
+        metadata = {k: v for k, v in example.items() if k not in [self.context_key, self.label_key]}
```

Comment:
Not sure if I'm doing something wrong, but I just tried this branch today and getting errors. I think the reason is that you only changed the `gpt_sft_dataset.py`, but I am using the `gpt_sft_chat_dataset.py`. Still the validation/inference part is shared for them and I'm getting errors, e.g. that metadata key is not available.

Am I doing something incorrectly or are the "chat"-based datasets broken in this branch @hsiehjackson?

Reply:
Did you see the error because of this? Let me add the check for metadata. https://github.com/NVIDIA/NeMo/blob/fix-gpt-trainer-test/nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py#L407

Reply:
that's just the first error - I fixed it in my code, but then everything breaks downstream as well. So I think more changes are needed here

Reply:
just confirmed that if I switch to the non-chat format, everything works well. I think you just need to make sure the dataset is prepared in the same way there including metadata, context, length, etc.

And as a side note - seems like we don't have any tests for chat-datasets? Something we should look into in the future, I guess

Reply:
`GPTSFTChatDataset` is fixed. Wait for @Kipok to try.